File: 0538_med_hot_convert-bst-to-greater-tree/solution.java
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * Definition for a binary tree node.
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
 
class Solution {
    public TreeNode convertBST(TreeNode root) {
        List<TreeNode> nodes = inorder(root);

        int size = nodes.size();

        for(int i = size - 1; i > 0; i--) {
            nodes.get(i-1).val += nodes.get(i).val;
        }

        return root;
    }

    private List<TreeNode> inorder(TreeNode root) {
        if(root == null) {
            return new ArrayList<>();
        }
        List<TreeNode> result = new ArrayList<>();

        result.addAll(inorder(root.left));
        result.add(root);
        result.addAll(inorder(root.right));

        return result;
    }

}

// 递归法，反中序遍历，右中左
class Solution2 {
    int addon = 0;

    public TreeNode convertBST(TreeNode root) {
        if (root == null) {
            return null;
        }

        inorder(root);

        return root;
    }

    private void inorder(TreeNode root) {
        if(root == null) {
            return;
        }

        inorder(root.right);

        root.val += addon;
        addon = root.val;

        inorder(root.left);
    }

}


class Solution3 {
    int addon = 0;

    public TreeNode convertBST(TreeNode root) {
        if (root == null) {
            return null;
        }

        inorder(root);

        return root;
    }

    private void inorder(TreeNode root) {
        if(root == null) {
            return;
        }

        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode cur = root;

        while(cur != null || !stack.isEmpty()) {
            while (cur != null) {
                stack.push(cur);
                cur = cur.right;
            }

            cur = stack.pop();

            cur.val += addon;
            addon = cur.val;

            cur = cur.left;
        }
    }

}

File: 0538_med_hot_convert-bst-to-greater-tree/solution.py
# Definition for a binary tree node.
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 迭代法，反中序遍历，采用右中左的遍历顺序
class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        stack = deque()

        addon = 0
        cur = root
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.right
            
            cur = stack.pop()
            cur.val += addon
            addon = cur.val

            cur = cur.left

        return root
        

# 先把二叉搜索树按中序遍历转为数组，然后在数组上处理
class Solution2:

    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        nodes = self.inorder(root)

        for i in range(len(nodes) - 1, 0, -1):
            nodes[i-1].val += nodes[i].val
        
        return root

    def inorder(self, root: Optional[TreeNode]) -> List[TreeNode]:
        if not root:
            return []
        
        return self.inorder(root.left) + [root] + self.inorder(root.right)
        

# 直接采用反中序的遍历方法来处理，边遍历边处理
class Solution3:
    def __init__(self) -> None:
        self.addon = 0

    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        self.inorder(root)

        return root

    def inorder(self, root: Optional[TreeNode]) -> None:
        if not root:
            return
        
        self.inorder(root.right)

        root.val += self.addon
        self.addon = root.val

        self.inorder(root.left)
        